photon-ml/src/main/scala/com/linkedin/photon/ml/model/MatrixFactorizationModel.scala

```diff
@@ -0,0 +1,107 @@
+    stringBuilder.append(s"\ncolLatentFactors L2 norm: $colLatentFactorsL2NormStats")
+    stringBuilder.toString()
+  }
+
+  override def equals(that: Any): Boolean = {
+    that match {
+      case other: MatrixFactorizationModel =>
+        val sameMetaData = this.rowEffectType == other.rowEffectType && this.colEffectType == other.colEffectType &&
+          this.numLatentFactors == other.numLatentFactors
+        val sameRowLatentFactors = this.rowLatentFactors.fullOuterJoin(other.rowLatentFactors).mapPartitions(iterator =>
+          Iterator.single(iterator.forall { case (_, (rowLatentFactor1, rowLatentFactor2)) =>
+            rowLatentFactor1.isDefined && rowLatentFactor2.isDefined &&
+              rowLatentFactor1.get.equals(rowLatentFactor2.get)
+          })
+        ).filter(!_).count() == 0
+        val sameColLatentFactors = this.colLatentFactors.fullOuterJoin(other.colLatentFactors).mapPartitions(iterator =>
+          Iterator.single(iterator.forall { case (_, (colLatentFactor1, colLatentFactor2)) =>
+            colLatentFactor1.isDefined && colLatentFactor2.isDefined &&
+              colLatentFactor1.get.equals(colLatentFactor2.get)
+          })
+        ).filter(!_).count() == 0
+        sameMetaData && sameRowLatentFactors && sameColLatentFactors
+      case _ => false
+    }
+  }
+}
+
+object MatrixFactorizationModel {
+  protected def score(
+    dataPoints: RDD[(Long, GameDatum)],
+    rowEffectType: String,
+    colEffectType: String,
+    rowLatentFactors: RDD[(String, Vector[Double])],
+    colLatentFactors: RDD[(String, Vector[Double])]): KeyValueScore = {
+    val scores = dataPoints.map { case (uniqueId, gameData) =>
+      val rowEffectId = gameData.randomEffectIdToIndividualIdMap(rowEffectType)
+      val colEffectId = gameData.randomEffectIdToIndividualIdMap(colEffectType)
+      (rowEffectId, (colEffectId, uniqueId))
+    }.cogroup(rowLatentFactors).flatMap { case (rowEffectId, (colEffectIdAndUniqueIdsIterable, rowLatentFactorIterable)) =>
+      assert(rowLatentFactorIterable.size <= 1, s"More than one row latent factor (${rowLatentFactorIterable.size}) " +
+        s"found for random effect Id $rowEffectId of random effect type $rowEffectType")
+      colEffectIdAndUniqueIdsIterable.flatMap { case (colEffectId, uniqueId) =>
+        rowLatentFactorIterable.map(rowLatentFactor => (colEffectId, (rowLatentFactor, uniqueId)))
+      }
+    }.cogroup(colLatentFactors).flatMap { case (colEffectId, (rowLatentFactorAndUniqueIdsIterable, colLatentFactorIterable)) =>
+      assert(colLatentFactorIterable.size <= 1, s"More than one column latent factor (${colLatentFactorIterable.size}) " +
+        s"found for random effect Id $colEffectId of random effect type $colEffectType")
+      rowLatentFactorAndUniqueIdsIterable.flatMap { case (rowLatentFactor, uniqueId) =>
+        colLatentFactorIterable.map(colLatentFactor => (uniqueId, rowLatentFactor.dot(colLatentFactor)) )
+      }
+    }
```

Comment:
Minor organizational note: when I read this, it looks like 4 stages to me:

1. For each datum, collect a ```(rowEffectId, (colEffectId, uniqueId))``` tuple.
2. Decorate rowEffectId with row latent factors.
3. Decorate colEffectId with column latent factors.
4. Compute dot product for ```(row latent factors, column latent factors)``` tuple.

You mostly have the code organized this way, but stages 3 & 4 are combined into one, which impedes readability a bit -- when I'm scanning the second ```cogroup``` block, my eye expects it to be a simple decoration step like the one above. I think it would help to separate the dot product computation into a separate ```map``` step after the two decoration steps. What do you think?

Also, I believe a brief one-liner comment above each stage would make this easier to parse.